src/hotspot/share/memory/metaspaceCriticalAllocation.cpp

```diff
@@ -113,19 +113,25 @@ void MetaspaceCriticalAllocation::remove(MetadataAllocationRequest* request) {
 }
 
 bool MetaspaceCriticalAllocation::try_allocate_critical(MetadataAllocationRequest* request) {
-  MutexLocker ml(MetaspaceCritical_lock, Mutex::_no_safepoint_check_flag);
-  if (_requests_head == request) {
-    // The first request can't opportunistically ride on a previous GC
-    return false;
+  {
+    MutexLocker ml(MetaspaceCritical_lock, Mutex::_no_safepoint_check_flag);
+    if (_requests_head == request) {
+      // The first request can't opportunistically ride on a previous GC
+      return false;
+    }
   }
   // Try to ride on a previous GC and hope for early satisfaction
   wait_for_purge(request);
   return request->result() != NULL;
 }
 
 void MetaspaceCriticalAllocation::wait_for_purge(MetadataAllocationRequest* request) {
-  while (!request->has_result()) {
+  for (;;) {
     ThreadBlockInVM tbivm(JavaThread::current());
```

Comment:
Can't you move the TBIVM outside of the loop now that it is always created?

Reply:
> Just a comment but it always concerns me that if we have to manually add a TBIVM when using a non-safepoint-checking lock then the lock is mis-classified as a non-safepoint-checking one! :(
> 
> That aside changes look fine. A few grammatical nits in the test.
> 
> Thanks, David

Thanks for the review David. I fixed your nits.

I agree that this lock should preferably have been a safepoint checking lock. In fact, it *was* a safepoint checking lock when I wrote the code. But that was before we changed the locking rules so that whether we do safepoint checking or not is a function of the rank. After that, this lock has become constrained to the current low rank by the current set of other locks, and by being that low ish rank, it is not allowed to safepoint check, unless I move a bunch of other lock ranks around, and figure out if it's okay for those locks to start safepoint checking as well, which isn't entirely obvious.

So this lock might be a case where those new rules end up being a bit awkward, and have lead this code to do manual transitions to blocked instead, as an escape hatch from the asserts.